### PR TITLE
[gardenlet] Introduce `controller-runtime` manager

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity scalability
/kind enhancement

**What this PR does / why we need it**:
Introduce `controller-runtime` manager in `gardenlet`.
No controllers have been refactored yet.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/4251

**Special notes for your reviewer**:
Similar to #6333.

**Release note**:

```breaking operator
`gardenlet`s component configuration API has been changed in the following breaking ways:
- `.server.http` has been split into `server.{healthProbes,metrics}` (health endpoints and metrics are now served on different ports)
- `.server.https` has been removed
```
```breaking operator
`gardenlet` serves health endpoints and metrics on different ports now. Adapt your scrape configs accordingly to port `metrics`.
```

### CODE REVIEW

File: pkg/client/kubernetes/client.go [hunk 1]
```diff
@@ -363,35 +363,35 @@ func defaultContentTypeProtobuf(c *rest.Config) *rest.Config {
 	return &config
 }
 
-var _ client.Client = &fallbackClient{}
+var _ client.Client = &FallbackClient{}
 
-// fallbackClient holds a `client.Reader` and `client.Reader` which is meant as a fallback
+// FallbackClient holds a `client.Reader` and `client.Reader` which is meant as a fallback
```
Comment (acumino) [inline pkg/client/kubernetes/client.go]: This doc string is confusing.

File: cmd/gardenlet/app/app.go [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+	if err != nil {
+		return err
+	}
+
+	log.Info("Setting up manager")
+	mgr, err := manager.New(seedRESTConfig, manager.Options{
+		Logger:                  log,
+		Scheme:                  kubernetes.SeedScheme,
+		HealthProbeBindAddress:  fmt.Sprintf("%s:%d", cfg.Server.HealthProbes.BindAddress, cfg.Server.HealthProbes.Port),
+		MetricsBindAddress:      fmt.Sprintf("%s:%d", cfg.Server.Metrics.BindAddress, cfg.Server.Metrics.Port),
```
Comment (acumino) [inline cmd/gardenlet/app/app.go:151]: Doc needs to be adapted similar to this https://github.com/gardener/gardener/pull/6624

File: charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl [hunk 3]
```diff
@@ -224,7 +224,8 @@ config.yaml: |
     {{- if .Values.global.gardenlet.config.leaderElection.resourceNamespace }}
     resourceNamespace: {{ .Values.global.gardenlet.config.leaderElection.resourceNamespace }}
     {{- end }}
-  logLevel: {{ required ".Values.global.gardenlet.config.logLevel is required" .Values.global.gardenlet.config.logLevel }}
+  logLevel: {{ .Values.global.gardenlet.config.logLevel }}
+  logFormat: {{ .Values.global.gardenlet.config.logFormat }}
```
Comment (oliver-goetz) [inline charts/gardener/gardenlet/charts/runtime/templates/_helpers.tpl:222]: There is no `logFormat` property in `charts/gardener/gardenlet/values.yaml` and `charts/gardener/gardenlet/charts/runtime/values.yaml`. They should be added here too to document all properties.

File: pkg/gardenlet/apis/config/v1alpha1/defaults.go [hunk 4]
```diff
@@ -60,14 +60,18 @@ func SetDefaults_GardenletConfiguration(obj *GardenletConfiguration) {
...
 
-	if obj.Server == nil {
-		obj.Server = &ServerConfiguration{}
+	if obj.Server.HealthProbes == nil {
+		obj.Server.HealthProbes = &Server{}
 	}
-	if len(obj.Server.HTTPS.BindAddress) == 0 {
-		obj.Server.HTTPS.BindAddress = "0.0.0.0"
+	if obj.Server.HealthProbes.Port == 0 {
+		obj.Server.HealthProbes.Port = 2728
```
Comment (oliver-goetz) [inline pkg/gardenlet/apis/config/v1alpha1/defaults.go:67]: Why is there a default for `Port`, but not for `BindAddress`? In the Helm charts before, both properties are marked as `required`.

File: pkg/gardenlet/apis/config/v1alpha1/defaults.go [hunk 5]
```diff
@@ -60,14 +60,18 @@ func SetDefaults_GardenletConfiguration(obj *GardenletConfiguration) {
...
 	}
-	if obj.Server.HTTPS.Port == 0 {
-		obj.Server.HTTPS.Port = 2720
+
+	if obj.Server.Metrics == nil {
+		obj.Server.Metrics = &Server{}
+	}
+	if obj.Server.Metrics.Port == 0 {
+		obj.Server.Metrics.Port = 2729
```
Comment (oliver-goetz) [inline pkg/gardenlet/apis/config/v1alpha1/defaults.go:74]: Same question as before.

**Review by oliver-goetz — COMMENTED** [review]
I went through it until commit `Make kubernetes.FallbackClient public` and found only minor things so far.

File: pkg/gardenlet/apis/config/v1alpha1/defaults.go (hunk 4 above)
Comment (rfranzke) [reply pkg/gardenlet/apis/config/v1alpha1/defaults.go:67]: I don't know, I copied this part from the `controller-manager` API to have it in sync: https://github.com/gardener/gardener/blob/1c7a68fb2e8208d6ad6f7eb88b6bca5345d5e1af/pkg/controllermanager/apis/config/v1alpha1/defaults.go#L202-L214
It is also done similarly for `gardener-admission-controller`: https://github.com/gardener/gardener/blob/1c7a68fb2e8208d6ad6f7eb88b6bca5345d5e1af/pkg/admissioncontroller/apis/config/v1alpha1/defaults.go#L46-L58

However, I agree that it's somewhat odd (from API point of view). The controller-runtime implicitly defaults an empty bind address to "0.0.0.0", I believe.

**Review by acumino — COMMENTED** [review]
/lgtm

**Review by oliver-goetz — APPROVED** [review]
/lgtm

**Pull request merged**